packages/@tailwindcss-upgrade/src/template/codemods/variant-order.ts

```diff
@@ -0,0 +1,125 @@
+import { walk, type AstNode } from '../../../../tailwindcss/src/ast'
+import type { Variant } from '../../../../tailwindcss/src/candidate'
+import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
+import { printCandidate } from '../candidates'
+
+export function variantOrder(designSystem: DesignSystem, rawCandidate: string): string {
+  for (let candidate of designSystem.parseCandidate(rawCandidate)) {
+    if (candidate.variants.length <= 1) {
+      continue
+    }
+
+    let mediaVariants = []
+    let regularVariants = []
+    let pseudoElementVariants = []
+
+    let originalOrder = candidate.variants
+
+    for (let variant of candidate.variants) {
+      if (isMediaVariant(designSystem, variant)) {
+        mediaVariants.push(variant)
+      } else if (isEndOfSelectorPseudoElement(variant)) {
+        pseudoElementVariants.push(variant)
+      } else {
+        regularVariants.push(variant)
+      }
+    }
+
+    // We only need to reorder regular variants if order is important
+    let regularVariantsNeedReordering = regularVariants.some((v) =>
+      isCombinatorVariant(designSystem, v),
+    )
+
+    // The candidate list in the AST need to be in reverse order
+    let newOrder = [
+      ...pseudoElementVariants,
+      ...(regularVariantsNeedReordering ? regularVariants.reverse() : regularVariants),
+      ...mediaVariants,
+    ]
+
+    if (orderMatches(originalOrder, newOrder)) {
+      continue
+    }
+
+    return printCandidate({ ...candidate, variants: newOrder })
+  }
+  return rawCandidate
+}
+
+function isMediaVariant(designSystem: DesignSystem, variant: Variant) {
+  // Handle the dark variant as a media variant
+  if (variant.kind === 'static' && variant.root === 'dark') {
+    return true
+  }
+  let stack = getAppliedNodeStack(designSystem, variant)
+  return stack.every((node) => node.kind === 'rule' && node.selector.startsWith('@media'))
+}
+
+function isCombinatorVariant(designSystem: DesignSystem, variant: Variant) {
+  let stack = getAppliedNodeStack(designSystem, variant)
+  return stack.some(
+    (node) =>
+      node.kind === 'rule' &&
+      // Ignore @media rules as they are hoisted
+      !node.selector.startsWith('@media') &&
+      // Combinators include any of the following characters
+      (node.selector.includes(' ') ||
+        node.selector.includes('>') ||
+        node.selector.includes('+') ||
+        node.selector.includes('~')),
+  )
+}
+
+function isEndOfSelectorPseudoElement(variant: Variant) {
+  if (variant.kind !== 'static') {
+    return false
+  }
+  switch (variant.root) {
+    case 'after':
+    case 'backdrop':
+    case 'before':
+    case 'first-letter':
+    case 'first-line':
+    case 'marker':
+    case 'placeholder':
+    case 'selection':
+      return true
+    default:
+      return false
+  }
```

Comment:
Instead of hardcoding specific utility names, this could also look at the generated CSS and check for these pseude-element selectors (similar to v3 where we fixed this by operating on the generated CSS)

Reply:
I think that'd be preferable because then we'd migrate custom variants correctly too.